### PR TITLE
Resolves #59: support backup to multiple storages

### LINKED ISSUE #59: How to dump once and save to multiple places?
I want to backup mysql, and the yml config just has 1 `store_with` i can set.

How to dump once and save to `local` and `s3`?

I don't want to set 2 models in config file.

Thank you.

### PR DESCRIPTION
**Pull request by ZachCheung**
* enable `storages` option
* backward compatible with `store_with`
* dedicated cycler config file for each backup storage

```yaml
    store_with:
      type: local
      keep: 1
      path: /tmp/demo
    storages:
      local1:
        type: local
        keep: 2
        path: /tmp/demo1
      s3:
        type: s3
        keep: 3
        bucket: wasabi2022
        region: ap-northeast-1
        endpoint: s3.ap-northeast-1.wasabisys.com
        path: backups/demo
        access_key_id: xxxxxxxxxxxxxxxxxxxx
        secret_access_key: xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
```

```
2022/12/01 19:45:14 [Modal: demo] WorkDir: /tmp/gobackup/1669952714853234082/demo
2022/12/01 19:45:14 [Database] => database | postgresql : postgresql
2022/12/01 19:45:14 [PostgreSQL] -> Dumping PostgreSQL...
2022/12/01 19:45:14 [PostgreSQL] dump path: /tmp/gobackup/1669952714853234082/demo/postgresql/postgresql/demo.sql
2022/12/01 19:45:14 [Database]
2022/12/01 19:45:14 [Compressor] => Compress | xz
2022/12/01 19:45:15 [Compressor] -> /tmp/gobackup/1669952714853234082/2022.12.01.19.45.14.tar.xz
2022/12/01 19:45:15 [Storage] => Storage | local
2022/12/01 19:45:15 [Local] Store successed /tmp/demo
2022/12/01 19:45:15 [Storage] => Storage | local
2022/12/01 19:45:15 [Local] Store successed /tmp/demo1
2022/12/01 19:45:15 [Storage] => Storage | s3
2022/12/01 19:45:15 [AWS S3] -> Uploading (0 MiB)...
2022/12/01 19:45:16 [AWS S3] => https://s3.ap-northeast-1.wasabisys.com/wasabi2022/backups/demo/2022.12.01.19.45.14.tar.xz
2022/12/01 19:45:16 [AWS S3] => s3://wasabi2022/backups/demo/2022.12.01.19.45.14.tar.xz
2022/12/01 19:45:16 [AWS S3] Duration 1 second 502 milliseconds, rate 1.0 MiB/s
2022/12/01 19:45:16 [Model] Cleanup temp: /tmp/gobackup/1669952714853234082/
```

```
2022/12/01 19:45:20 [Modal: demo] WorkDir: /tmp/gobackup/1669952720652119244/demo
2022/12/01 19:45:20 [Database] => database | postgresql : postgresql
2022/12/01 19:45:20 [PostgreSQL] -> Dumping PostgreSQL...
2022/12/01 19:45:20 [PostgreSQL] dump path: /tmp/gobackup/1669952720652119244/demo/postgresql/postgresql/demo.sql
2022/12/01 19:45:20 [Database]
2022/12/01 19:45:20 [Compressor] => Compress | xz
2022/12/01 19:45:20 [Compressor] -> /tmp/gobackup/1669952720652119244/2022.12.01.19.45.20.tar.xz
2022/12/01 19:45:20 [Storage] => Storage | local
2022/12/01 19:45:20 [Local] Store successed /tmp/demo
2022/12/01 19:45:20 [Cycler] Removed 2022.12.01.19.45.14.tar.xz
2022/12/01 19:45:20 [Storage] => Storage | local
2022/12/01 19:45:20 [Local] Store successed /tmp/demo1
2022/12/01 19:45:20 [Storage] => Storage | s3
2022/12/01 19:45:20 [AWS S3] -> Uploading (0 MiB)...
2022/12/01 19:45:21 [AWS S3] => https://s3.ap-northeast-1.wasabisys.com/wasabi2022/backups/demo/2022.12.01.19.45.20.tar.xz
2022/12/01 19:45:21 [AWS S3] => s3://wasabi2022/backups/demo/2022.12.01.19.45.20.tar.xz
2022/12/01 19:45:21 [AWS S3] Duration 608 milliseconds 124 microseconds, rate 1.0 MiB/s
2022/12/01 19:45:21 [Model] Cleanup temp: /tmp/gobackup/1669952720652119244/
```

```
2022/12/01 19:45:26 [Modal: demo] WorkDir: /tmp/gobackup/1669952726786464912/demo
2022/12/01 19:45:26 [Database] => database | postgresql : postgresql
2022/12/01 19:45:26 [PostgreSQL] -> Dumping PostgreSQL...
2022/12/01 19:45:26 [PostgreSQL] dump path: /tmp/gobackup/1669952726786464912/demo/postgresql/postgresql/demo.sql
2022/12/01 19:45:26 [Database]
2022/12/01 19:45:26 [Compressor] => Compress | xz
2022/12/01 19:45:26 [Compressor] -> /tmp/gobackup/1669952726786464912/2022.12.01.19.45.26.tar.xz
2022/12/01 19:45:26 [Storage] => Storage | local
2022/12/01 19:45:26 [Local] Store successed /tmp/demo
2022/12/01 19:45:26 [Cycler] Removed 2022.12.01.19.45.20.tar.xz
2022/12/01 19:45:26 [Storage] => Storage | local
2022/12/01 19:45:26 [Local] Store successed /tmp/demo1
2022/12/01 19:45:26 [Cycler] Removed 2022.12.01.19.45.14.tar.xz
2022/12/01 19:45:26 [Storage] => Storage | s3
2022/12/01 19:45:26 [AWS S3] -> Uploading (0 MiB)...
2022/12/01 19:45:27 [AWS S3] => https://s3.ap-northeast-1.wasabisys.com/wasabi2022/backups/demo/2022.12.01.19.45.26.tar.xz
2022/12/01 19:45:27 [AWS S3] => s3://wasabi2022/backups/demo/2022.12.01.19.45.26.tar.xz
2022/12/01 19:45:27 [AWS S3] Duration 535 milliseconds 185 microseconds, rate 1.0 MiB/s
2022/12/01 19:45:27 [Model] Cleanup temp: /tmp/gobackup/1669952726786464912/
```

### CODE REVIEW

**Comment by huacnlee** [comment]
How about:

```yml
store_with:
  type: multiple
  storages:
    local1:
      type: local
      keep: 2
      path: /tmp/demo
    s3:
      type: s3
      keep: 3
      bucket: wasabi2022
      region: ap-northeast-1
      endpoint: s3.ap-northeast-1.wasabisys.com
      path: backups/demo
      access_key_id: xxxxxxxxxxxxxxxxxxxx
      secret_access_key: xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
```

This means add a `multiple` type for `store_with`, so we can keep supports for old config rules.

**Comment by ZachCheung** [comment]
```yaml
store_with:
  type: local
  keep: 1
  path: /tmp/demo
storages:
  local1:
    type: local
    keep: 2
    path: /tmp/demo1
```

It supports the old config rules actually. The current way is merging `store_with` into `storages` logically and if user only needs one storage then it's okay to stick with `store_with` meanwhile if user desires multiple storages then `storages` should be configured.

```yaml
store_with:
  type: multiple
  storages:
    local1:
      type: local
      keep: 2
      path: /tmp/demo
    s3:
      type: s3
      keep: 3
      bucket: wasabi2022
      region: ap-northeast-1
      endpoint: s3.ap-northeast-1.wasabisys.com
      path: backups/demo
      access_key_id: xxxxxxxxxxxxxxxxxxxx
      secret_access_key: xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
```

```yaml
store_with:
  type: local
  keep: 1
  path: /tmp/demo
```

If we add a `multiple` type for `store_with` then `store_with` has two config structures which look quite different.

The previous way is like `store_with` is still supported but somehow deprecated and `storages` is required for multiple storages, while the latter one is introducing a new config structure of `store_with` for multiple storages support. I'm not quite sure which way is better, what do you think?

**Comment by huacnlee** [comment]
Okey, let change to use `storages`, to same as the `databases`.

And you need add `Deprecated` warning, when use still use `store_with`

**Comment by ZachCheung** [comment]
```
2022/12/02 00:37:51 [DeprecationWarning] "store_with" is deprecated now, please use "storages" which supports multiple storages.
2022/12/02 00:37:51 [Modal: demo] WorkDir: /tmp/gobackup/1669970271079006943/demo
2022/12/02 00:37:51 [Database] => database | postgresql : postgresql
2022/12/02 00:37:51 [PostgreSQL] -> Dumping PostgreSQL...
2022/12/02 00:37:51 [PostgreSQL] dump path: /tmp/gobackup/1669970271079006943/demo/postgresql/postgresql/demo.sql
2022/12/02 00:37:51 [Database]
2022/12/02 00:37:51 [Compressor] => Compress | xz
2022/12/02 00:37:51 [Compressor] -> /tmp/gobackup/1669970271079006943/2022.12.02.00.37.51.tar.xz
2022/12/02 00:37:51 [Storage] => Storage | local
2022/12/02 00:37:51 [Local] Store successed /tmp/demo
2022/12/02 00:37:51 [Cycler] Removed 2022.12.02.00.36.19.tar.xz
2022/12/02 00:37:51 [Model] Cleanup temp: /tmp/gobackup/1669970271079006943/
```

File: config/config.go [hunk 1]
```diff
@@ -115,6 +115,7 @@ func loadStoragesConfig(model *ModelConfig) {
 	// Backward compatible with `store_with` config
 	storeWith := model.Viper.Sub("store_with")
 	if storeWith != nil {
+		logger.Warn(`[DeprecationWarning] "store_with" is deprecated now, please use "storages" which supports multiple storages.`)
```
Comment (huacnlee) [inline config/config.go:118]: `[Deprecated]`

**Comment by ZachCheung** [comment]
Saw the failed tests, fixing.

**Comment by huacnlee** [comment]
I just fixed test.

**Comment by ZachCheung** [comment]
Cool.

**Pull request merged**